File: packages/ts/cairn-p2p/src/crypto/noise.ts
import { sha256 } from '@noble/hashes/sha256';
import { ed25519 } from '@noble/curves/ed25519';
import { x25519 } from '@noble/curves/ed25519';
import { CairnError } from '../errors.js';
import { hkdfSha256, HKDF_INFO_SESSION_KEY } from './hkdf.js';
import { aeadEncrypt, aeadDecrypt } from './aead.js';
import type { IdentityKeypair } from './identity.js';

/** Protocol name used to initialize the handshake hash (Noise spec section 5.2). */
const PROTOCOL_NAME = new TextEncoder().encode('Noise_XX_25519_ChaChaPoly_SHA256');

/** AEAD tag size (16 bytes for ChaCha20-Poly1305). */
const TAG_SIZE = 16;
/** Size of an X25519 public key. */
const DH_KEY_SIZE = 32;
/** Size of an Ed25519 public key. */
const ED25519_PUB_SIZE = 32;
/** Zero nonce used for handshake AEAD operations. */
const ZERO_NONCE = new Uint8Array(12);

/** Noise XX handshake role. */
export type Role = 'initiator' | 'responder';

/** Result of a completed Noise XX handshake. */
export interface HandshakeResult {
  /** Shared symmetric key for session encryption (32 bytes). */
  sessionKey: Uint8Array;
  /** Remote peer's static public key (Ed25519, 32 bytes). */
  remoteStatic: Uint8Array;
  /** Handshake transcript hash for SAS derivation (32 bytes). */
  transcriptHash: Uint8Array;
}

/** Output of a single handshake step. */
export type StepOutput =
  | { type: 'send_message'; data: Uint8Array }
  | { type: 'complete'; result: HandshakeResult };

/** Internal handshake state. */
type HandshakeState =
  | 'initiator_start'
  | 'responder_wait_msg1'
  | 'initiator_wait_msg2'
  | 'responder_wait_msg3'
  | 'complete';

/**
 * Noise XX handshake state machine.
 *
 * Implements the three-message Noise XX pattern:
 * ```
 * -> e                 (message 1)
 * <- e, ee, s, es      (message 2)
 * -> s, se             (message 3)
 * ```
 *
 * Initiator flow:
 *   1. step()        -> SendMessage(msg1)
 *   2. step(msg2)    -> SendMessage(msg3) — call result() to get HandshakeResult
 *   3. (done — call result() after step 2)
 *
 * Responder flow:
 *   1. step(msg1)    -> SendMessage(msg2)
 *   2. step(msg3)    -> Complete(result)
 */
export class NoiseXXHandshake {
  private state: HandshakeState;
  private readonly localIdentity: IdentityKeypair;
  private readonly localStaticX25519Secret: Uint8Array;

  private localEphemeralSecret?: Uint8Array;
  private localEphemeralPub?: Uint8Array;
  private remoteEphemeralPub?: Uint8Array;
  private remoteStaticEd25519?: Uint8Array;

  private chainingKey: Uint8Array;
  private handshakeHash: Uint8Array;
  private currentKey?: Uint8Array;
  private pakeSecret?: Uint8Array;
  private cachedResult?: HandshakeResult;

  constructor(role: Role, identity: IdentityKeypair, pakeSecret?: Uint8Array) {
    // Convert Ed25519 identity to X25519 for DH operations.
    this.localStaticX25519Secret = ed25519.utils.toMontgomerySecret(identity.secretBytes());

    // Initialize handshake hash from protocol name (Noise spec section 5.2).
    // If protocol name <= 32 bytes, pad with zeros.
    if (PROTOCOL_NAME.length <= 32) {
      this.handshakeHash = new Uint8Array(32);
      this.handshakeHash.set(PROTOCOL_NAME);
    } else {
      this.handshakeHash = sha256(PROTOCOL_NAME);
    }

    // Chaining key starts as the handshake hash (Noise spec).
    this.chainingKey = new Uint8Array(this.handshakeHash);

    this.localIdentity = identity;
    this.state = role === 'initiator' ? 'initiator_start' : 'responder_wait_msg1';

    if (pakeSecret) {
      if (pakeSecret.length !== 32) {
        throw new CairnError('CRYPTO', 'PAKE secret must be 32 bytes');
      }
      this.pakeSecret = new Uint8Array(pakeSecret);
    }
  }

  /**
   * Process the next handshake step.
   *
   * Initiator: step() -> step(msg2)
   * Responder: step(msg1) -> step(msg3)
   */
  step(input?: Uint8Array): StepOutput {
    switch (this.state) {
      case 'initiator_start':
        if (input !== undefined) {
          throw new CairnError('CRYPTO', 'initiator start expects no input');
        }
        return this.initiatorSendMsg1();

      case 'responder_wait_msg1':
        if (input === undefined) {
          throw new CairnError('CRYPTO', 'responder expects message 1 input');
        }
        return this.responderRecvMsg1SendMsg2(input);

      case 'initiator_wait_msg2':
        if (input === undefined) {
          throw new CairnError('CRYPTO', 'initiator expects message 2 input');
        }
        return this.initiatorRecvMsg2SendMsg3(input);

      case 'responder_wait_msg3':
        if (input === undefined) {
          throw new CairnError('CRYPTO', 'responder expects message 3 input');
        }
        return this.responderRecvMsg3(input);

      case 'complete':
        throw new CairnError('CRYPTO', 'handshake already complete');
    }
  }

  /** Get the handshake result after the initiator has sent message 3. */
  getResult(): HandshakeResult {
    if (!this.cachedResult) {
      throw new CairnError('CRYPTO', 'handshake not yet complete');
    }
    return this.cachedResult;
  }

  // --- Message 1: -> e ---

  private initiatorSendMsg1(): StepOutput {
    // Generate ephemeral keypair
    const ephemeralSecret = x25519.utils.randomSecretKey();
    const ephemeralPub = x25519.getPublicKey(ephemeralSecret);

    // Mix ephemeral public key into handshake hash
    this.mixHash(ephemeralPub);

    this.localEphemeralSecret = ephemeralSecret;
    this.localEphemeralPub = ephemeralPub;

    // Message 1 is just the ephemeral public key (32 bytes)
    this.state = 'initiator_wait_msg2';
    return { type: 'send_message', data: new Uint8Array(ephemeralPub) };
  }

  // --- Message 2: <- e, ee, s, es ---

  private responderRecvMsg1SendMsg2(msg1: Uint8Array): StepOutput {
    if (msg1.length !== DH_KEY_SIZE) {
      throw new CairnError('CRYPTO', `message 1 invalid length: expected ${DH_KEY_SIZE}, got ${msg1.length}`);
    }

    // Store remote ephemeral and mix into hash
    this.remoteEphemeralPub = new Uint8Array(msg1);
    this.mixHash(this.remoteEphemeralPub);

    // Build message 2
    const parts: Uint8Array[] = [];

    // e: generate responder ephemeral
    const ephemeralSecret = x25519.utils.randomSecretKey();
    const ephemeralPub = x25519.getPublicKey(ephemeralSecret);

    this.mixHash(ephemeralPub);
    parts.push(ephemeralPub);

    this.localEphemeralSecret = ephemeralSecret;
    this.localEphemeralPub = ephemeralPub;

    // ee: DH(responder_ephemeral, initiator_ephemeral)
    const eeShared = x25519.getSharedSecret(ephemeralSecret, this.remoteEphemeralPub);
    this.mixKey(eeShared);

    // s: encrypt and send our static Ed25519 public key
    const staticPubBytes = this.localIdentity.publicKey();
    const encryptedStatic = this.encryptAndHash(staticPubBytes);
    parts.push(encryptedStatic);

    // es: DH(responder_static_x25519, initiator_ephemeral)
    const esShared = x25519.getSharedSecret(this.localStaticX25519Secret, this.remoteEphemeralPub);
    this.mixKey(esShared);

    // Encrypt empty payload
    const encryptedPayload = this.encryptAndHash(new Uint8Array(0));
    parts.push(encryptedPayload);

    this.state = 'responder_wait_msg3';
    return { type: 'send_message', data: concatBytes(...parts) };
  }

  // --- Initiator: recv message 2, send message 3 ---

  private initiatorRecvMsg2SendMsg3(msg2: Uint8Array): StepOutput {
    const minLen = DH_KEY_SIZE + (ED25519_PUB_SIZE + TAG_SIZE) + TAG_SIZE;
    if (msg2.length < minLen) {
      throw new CairnError('CRYPTO', `message 2 too short: expected at least ${minLen}, got ${msg2.length}`);
    }

    let offset = 0;

    // e: responder ephemeral
    const remoteEBytes = msg2.slice(offset, offset + DH_KEY_SIZE);
    this.mixHash(remoteEBytes);
    offset += DH_KEY_SIZE;
    this.remoteEphemeralPub = remoteEBytes;

    // ee: DH(initiator_ephemeral, responder_ephemeral)
    if (!this.localEphemeralSecret) {
      throw new CairnError('CRYPTO', 'missing local ephemeral for ee DH');
    }
    const eeShared = x25519.getSharedSecret(this.localEphemeralSecret, remoteEBytes);
    this.mixKey(eeShared);

    // s: decrypt responder's static public key
    const encryptedStatic = msg2.slice(offset, offset + ED25519_PUB_SIZE + TAG_SIZE);
    const staticPubBytes = this.decryptAndHash(encryptedStatic);
    offset += ED25519_PUB_SIZE + TAG_SIZE;

    if (staticPubBytes.length !== ED25519_PUB_SIZE) {
      throw new CairnError('CRYPTO', 'decrypted static key wrong size');
    }

    // Convert remote Ed25519 public key to X25519 for DH
    const remoteStaticX25519 = ed25519.utils.toMontgomery(staticPubBytes);
    this.remoteStaticEd25519 = new Uint8Array(staticPubBytes);

    // es: DH(initiator_ephemeral, responder_static_x25519)
    const esShared = x25519.getSharedSecret(this.localEphemeralSecret, remoteStaticX25519);
    this.mixKey(esShared);

    // Decrypt payload from message 2
    const encryptedPayload = msg2.slice(offset);
    this.decryptAndHash(encryptedPayload);

    // Build message 3: -> s, se
    const parts: Uint8Array[] = [];

    // s: encrypt initiator's static Ed25519 public key
    const ourStaticPubBytes = this.localIdentity.publicKey();
    const encryptedOurStatic = this.encryptAndHash(ourStaticPubBytes);
    parts.push(encryptedOurStatic);

    // se: DH(initiator_static_x25519, responder_ephemeral)
    const seShared = x25519.getSharedSecret(this.localStaticX25519Secret, remoteEBytes);
    this.mixKey(seShared);

    // Mix in PAKE secret if present
    if (this.pakeSecret) {
      this.mixKey(this.pakeSecret);
    }

    // Encrypt empty payload for message 3
    const encryptedMsg3Payload = this.encryptAndHash(new Uint8Array(0));
    parts.push(encryptedMsg3Payload);

    // Derive session key
    const sessionKey = this.deriveSessionKey();
    const result: HandshakeResult = {
      sessionKey,
      remoteStatic: new Uint8Array(staticPubBytes),
      transcriptHash: new Uint8Array(this.handshakeHash),
    };

    this.state = 'complete';
    this.cachedResult = result;

    return { type: 'send_message', data: concatBytes(...parts) };
  }

  // --- Message 3: responder receives -> s, se ---

  private responderRecvMsg3(msg3: Uint8Array): StepOutput {
    const minLen = (ED25519_PUB_SIZE + TAG_SIZE) + TAG_SIZE;
    if (msg3.length < minLen) {
      throw new CairnError('CRYPTO', `message 3 too short: expected at least ${minLen}, got ${msg3.length}`);
    }

    let offset = 0;

    // s: decrypt initiator's static public key
    const encryptedStatic = msg3.slice(offset, offset + ED25519_PUB_SIZE + TAG_SIZE);
    const staticPubBytes = this.decryptAndHash(encryptedStatic);
    offset += ED25519_PUB_SIZE + TAG_SIZE;

    if (staticPubBytes.length !== ED25519_PUB_SIZE) {
      throw new CairnError('CRYPTO', 'decrypted static key wrong size');
    }

    // Convert remote Ed25519 to X25519
    const remoteStaticX25519 = ed25519.utils.toMontgomery(staticPubBytes);
    this.remoteStaticEd25519 = new Uint8Array(staticPubBytes);

    // se: DH(responder_ephemeral, initiator_static_x25519)
    if (!this.localEphemeralSecret) {
      throw new CairnError('CRYPTO', 'missing local ephemeral for se DH');
    }
    const seShared = x25519.getSharedSecret(this.localEphemeralSecret, remoteStaticX25519);
    this.mixKey(seShared);

    // Mix in PAKE secret if present
    if (this.pakeSecret) {
      this.mixKey(this.pakeSecret);
    }

    // Decrypt payload
    const encryptedPayload = msg3.slice(offset);
    this.decryptAndHash(encryptedPayload);

    // Derive session key
    const sessionKey = this.deriveSessionKey();

    this.state = 'complete';
    const result: HandshakeResult = {
      sessionKey,
      remoteStatic: new Uint8Array(staticPubBytes),
      transcriptHash: new Uint8Array(this.handshakeHash),
    };
    this.cachedResult = result;
    return { type: 'complete', result };
  }

  // --- Noise symmetric state operations ---

  /**
   * Mix a DH result into the chaining key via HKDF.
   * Updates the chaining key and stores the derived encryption key.
   */
  private mixKey(inputKeyMaterial: Uint8Array): void {
    const output = hkdfSha256(inputKeyMaterial, this.chainingKey, new Uint8Array(0), 64);
    this.chainingKey = output.slice(0, 32);
    this.currentKey = output.slice(32, 64);
  }

  /** Mix data into the handshake hash: h = SHA-256(h || data). */
  private mixHash(data: Uint8Array): void {
    const combined = concatBytes(this.handshakeHash, data);
    this.handshakeHash = sha256(combined);
  }

  /** Encrypt plaintext and mix the ciphertext into the handshake hash. */
  private encryptAndHash(plaintext: Uint8Array): Uint8Array {
    if (!this.currentKey) {
      throw new CairnError('CRYPTO', 'no encryption key available (mixKey not called)');
    }
    const ciphertext = aeadEncrypt(
      'chacha20-poly1305',
      this.currentKey,
      ZERO_NONCE,
      plaintext,
      this.handshakeHash,
    );
    this.mixHash(ciphertext);
    return ciphertext;
  }

  /** Decrypt ciphertext and mix it into the handshake hash. */
  private decryptAndHash(ciphertext: Uint8Array): Uint8Array {
    if (!this.currentKey) {
      throw new CairnError('CRYPTO', 'no decryption key available (mixKey not called)');
    }
    const hBefore = new Uint8Array(this.handshakeHash);
    this.mixHash(ciphertext);
    return aeadDecrypt(
      'chacha20-poly1305',
      this.currentKey,
      ZERO_NONCE,
      ciphertext,
      hBefore,
    );
  }

  /** Derive the final session key from the chaining key. */
  private deriveSessionKey(): Uint8Array {
    return hkdfSha256(this.chainingKey, undefined, HKDF_INFO_SESSION_KEY, 32);
  }
}

/** Concatenate multiple Uint8Arrays. */
function concatBytes(...arrays: Uint8Array[]): Uint8Array {
  let totalLen = 0;
  for (const arr of arrays) totalLen += arr.length;
  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const arr of arrays) {
    result.set(arr, offset);
    offset += arr.length;
  }
  return result;
}

File: packages/go/cairn-p2p/crypto/noise.go
package crypto

import (
	"crypto/ed25519"
	"crypto/sha256"
	"crypto/sha512"
	"fmt"

	"filippo.io/edwards25519"
	"golang.org/x/crypto/curve25519"
)

// ProtocolName is the Noise protocol name used to initialize the handshake hash.
const ProtocolName = "Noise_XX_25519_ChaChaPoly_SHA256"

const (
	dhKeySize     = 32
	ed25519PubLen = 32
	tagSize       = 16 // ChaCha20-Poly1305 tag
)

var zeroNonce [12]byte

// EmojiTable is the 64-entry table for emoji SAS derivation.
var EmojiTable = [64]string{
	"dog", "cat", "fish", "bird", "bear", "lion", "wolf", "fox",
	"deer", "owl", "bee", "ant", "star", "moon", "sun", "fire",
	"tree", "leaf", "rose", "wave", "rain", "snow", "bolt", "wind",
	"rock", "gem", "bell", "key", "lock", "flag", "book", "pen",
	"cup", "hat", "shoe", "ring", "cake", "gift", "lamp", "gear",
	"ship", "car", "bike", "drum", "horn", "harp", "dice", "coin",
	"map", "tent", "crown", "sword", "shield", "bow", "axe", "hammer",
	"anchor", "wheel", "clock", "heart", "skull", "ghost", "robot", "alien",
}

// Role identifies the handshake role.
type Role int

const (
	RoleInitiator Role = iota
	RoleResponder
)

type handshakeState int

const (
	stateInitiatorStart handshakeState = iota
	stateResponderWaitMsg1
	stateInitiatorWaitMsg2
	stateResponderWaitMsg3
	stateComplete
)

// HandshakeResult is the output of a completed Noise XX handshake.
type HandshakeResult struct {
	SessionKey     [32]byte
	RemoteStatic   ed25519.PublicKey
	TranscriptHash [32]byte
}

// StepOutput is the output of a single handshake step.
type StepOutput struct {
	Message  []byte           // non-nil if we need to send
	Complete *HandshakeResult // non-nil if handshake finished
}

// NoiseXXHandshake implements the Noise XX pattern state machine.
type NoiseXXHandshake struct {
	role  Role
	state handshakeState

	localIdentity      *IdentityKeypair
	localStaticX25519  [32]byte // private key
	localEphemeral     *[32]byte
	localEphemeralPub  *[32]byte
	remoteEphemeral    *[32]byte
	remoteStatic       ed25519.PublicKey
	chainingKey        [32]byte
	handshakeHash      [32]byte
	currentKey         *[32]byte
	pakeSecret         *[32]byte
	cachedResult       *HandshakeResult
}

// NewNoiseXX creates a new Noise XX handshake state machine.
func NewNoiseXX(role Role, identity *IdentityKeypair, pakeSecret []byte) *NoiseXXHandshake {
	// Convert Ed25519 signing key to X25519 private key
	localStaticX25519 := ed25519PrivateToX25519(identity.privateKey)

	// Initialize handshake hash from protocol name (Noise spec section 5.2).
	// If protocol name <= 32 bytes, pad with zeros. If > 32, hash it.
	var handshakeHash [32]byte
	if len(ProtocolName) <= 32 {
		copy(handshakeHash[:], ProtocolName)
	} else {
		handshakeHash = sha256.Sum256([]byte(ProtocolName))
	}

	chainingKey := handshakeHash

	state := stateInitiatorStart
	if role == RoleResponder {
		state = stateResponderWaitMsg1
	}

	h := &NoiseXXHandshake{
		role:              role,
		state:             state,
		localIdentity:     identity,
		localStaticX25519: localStaticX25519,
		chainingKey:       chainingKey,
		handshakeHash:     handshakeHash,
	}

	if len(pakeSecret) == 32 {
		var ps [32]byte
		copy(ps[:], pakeSecret)
		h.pakeSecret = &ps
	}

	return h
}

// Step processes the next handshake step.
func (h *NoiseXXHandshake) Step(input []byte) (*StepOutput, error) {
	switch h.state {
	case stateInitiatorStart:
		if input != nil {
			return nil, fmt.Errorf("initiator start expects no input")
		}
		return h.initiatorSendMsg1()
	case stateResponderWaitMsg1:
		if input == nil {
			return nil, fmt.Errorf("responder expects message 1 input")
		}
		return h.responderRecvMsg1SendMsg2(input)
	case stateInitiatorWaitMsg2:
		if input == nil {
			return nil, fmt.Errorf("initiator expects message 2 input")
		}
		return h.initiatorRecvMsg2SendMsg3(input)
	case stateResponderWaitMsg3:
		if input == nil {
			return nil, fmt.Errorf("responder expects message 3 input")
		}
		return h.responderRecvMsg3(input)
	case stateComplete:
		return nil, fmt.Errorf("handshake already complete")
	default:
		return nil, fmt.Errorf("invalid handshake state")
	}
}

// Result returns the cached handshake result for the initiator after msg3 is sent.
func (h *NoiseXXHandshake) Result() (*HandshakeResult, error) {
	if h.cachedResult == nil {
		return nil, fmt.Errorf("handshake not yet complete")
	}
	return h.cachedResult, nil
}

// --- Message 1: -> e ---

func (h *NoiseXXHandshake) initiatorSendMsg1() (*StepOutput, error) {
	ephPriv, ephPub, err := generateX25519Keypair()
	if err != nil {
		return nil, err
	}

	h.mixHash(ephPub[:])
	h.localEphemeral = &ephPriv
	h.localEphemeralPub = &ephPub

	msg := make([]byte, 32)
	copy(msg, ephPub[:])

	h.state = stateInitiatorWaitMsg2
	return &StepOutput{Message: msg}, nil
}

// --- Message 2: <- e, ee, s, es ---

func (h *NoiseXXHandshake) responderRecvMsg1SendMsg2(msg1 []byte) (*StepOutput, error) {
	if len(msg1) != dhKeySize {
		return nil, fmt.Errorf("message 1 invalid length: expected %d, got %d", dhKeySize, len(msg1))
	}

	var remoteE [32]byte
	copy(remoteE[:], msg1)
	h.mixHash(remoteE[:])
	h.remoteEphemeral = &remoteE

	// e: generate responder ephemeral
	ephPriv, ephPub, err := generateX25519Keypair()
	if err != nil {
		return nil, err
	}
	h.mixHash(ephPub[:])
	h.localEphemeral = &ephPriv
	h.localEphemeralPub = &ephPub

	var msg2 []byte
	msg2 = append(msg2, ephPub[:]...)

	// ee: DH(responder_ephemeral, initiator_ephemeral)
	eeShared, err := x25519DH(ephPriv, remoteE)
	if err != nil {
		return nil, err
	}
	if err := h.mixKey(eeShared); err != nil {
		return nil, err
	}

	// s: encrypt and send our static Ed25519 public key
	staticPubBytes := []byte(h.localIdentity.PublicKey())
	encryptedStatic, err := h.encryptAndHash(staticPubBytes)
	if err != nil {
		return nil, err
	}
	msg2 = append(msg2, encryptedStatic...)

	// es: DH(responder_static_x25519, initiator_ephemeral)
	esShared, err := x25519DH(h.localStaticX25519, remoteE)
	if err != nil {
		return nil, err
	}
	if err := h.mixKey(esShared); err != nil {
		return nil, err
	}

	// Encrypt empty payload
	encryptedPayload, err := h.encryptAndHash(nil)
	if err != nil {
		return nil, err
	}
	msg2 = append(msg2, encryptedPayload...)

	h.state = stateResponderWaitMsg3
	return &StepOutput{Message: msg2}, nil
}

// --- Initiator: recv message 2, send message 3 ---

func (h *NoiseXXHandshake) initiatorRecvMsg2SendMsg3(msg2 []byte) (*StepOutput, error) {
	minLen := dhKeySize + (ed25519PubLen + tagSize) + tagSize
	if len(msg2) < minLen {
		return nil, fmt.Errorf("message 2 too short: expected at least %d, got %d", minLen, len(msg2))
	}

	offset := 0

	// e: responder ephemeral
	var remoteE [32]byte
	copy(remoteE[:], msg2[offset:offset+dhKeySize])
	h.mixHash(remoteE[:])
	offset += dhKeySize
	h.remoteEphemeral = &remoteE

	// ee: DH(initiator_ephemeral, responder_ephemeral)
	if h.localEphemeral == nil {
		return nil, fmt.Errorf("missing local ephemeral for ee DH")
	}
	eeShared, err := x25519DH(*h.localEphemeral, remoteE)
	if err != nil {
		return nil, err
	}
	if err := h.mixKey(eeShared); err != nil {
		return nil, err
	}

	// s: decrypt responder's static public key
	encryptedStatic := msg2[offset : offset+ed25519PubLen+tagSize]
	staticPubBytes, err := h.decryptAndHash(encryptedStatic)
	if err != nil {
		return nil, err
	}
	offset += ed25519PubLen + tagSize

	if len(staticPubBytes) != ed25519PubLen {
		return nil, fmt.Errorf("decrypted static key wrong size")
	}

	remoteStaticPub := ed25519.PublicKey(staticPubBytes)
	h.remoteStatic = remoteStaticPub

	// Convert remote Ed25519 public key to X25519 for DH
	remoteStaticX25519 := ed25519PublicToX25519(remoteStaticPub)

	// es: DH(initiator_ephemeral, responder_static_x25519)
	esShared, err := x25519DH(*h.localEphemeral, remoteStaticX25519)
	if err != nil {
		return nil, err
	}
	if err := h.mixKey(esShared); err != nil {
		return nil, err
	}

	// Decrypt payload from message 2
	encryptedPayload := msg2[offset:]
	if _, err := h.decryptAndHash(encryptedPayload); err != nil {
		return nil, err
	}

	// Now build message 3: -> s, se
	var msg3 []byte

	// s: encrypt initiator's static Ed25519 public key
	ourStaticPub := []byte(h.localIdentity.PublicKey())
	encryptedOurStatic, err := h.encryptAndHash(ourStaticPub)
	if err != nil {
		return nil, err
	}
	msg3 = append(msg3, encryptedOurStatic...)

	// se: DH(initiator_static_x25519, responder_ephemeral)
	seShared, err := x25519DH(h.localStaticX25519, remoteE)
	if err != nil {
		return nil, err
	}
	if err := h.mixKey(seShared); err != nil {
		return nil, err
	}

	// Mix in PAKE secret if present
	if h.pakeSecret != nil {
		if err := h.mixKey(*h.pakeSecret); err != nil {
			return nil, err
		}
	}

	// Encrypt empty payload for message 3
	encPayload, err := h.encryptAndHash(nil)
	if err != nil {
		return nil, err
	}
	msg3 = append(msg3, encPayload...)

	// Derive session key
	sessionKey, err := h.deriveSessionKey()
	if err != nil {
		return nil, err
	}

	h.state = stateComplete
	h.cachedResult = &HandshakeResult{
		SessionKey:     sessionKey,
		RemoteStatic:   remoteStaticPub,
		TranscriptHash: h.handshakeHash,
	}

	return &StepOutput{Message: msg3}, nil
}

// --- Message 3: responder receives -> s, se ---

func (h *NoiseXXHandshake) responderRecvMsg3(msg3 []byte) (*StepOutput, error) {
	minLen := (ed25519PubLen + tagSize) + tagSize
	if len(msg3) < minLen {
		return nil, fmt.Errorf("message 3 too short: expected at least %d, got %d", minLen, len(msg3))
	}

	offset := 0

	// s: decrypt initiator's static public key
	encryptedStatic := msg3[offset : offset+ed25519PubLen+tagSize]
	staticPubBytes, err := h.decryptAndHash(encryptedStatic)
	if err != nil {
		return nil, err
	}
	offset += ed25519PubLen + tagSize

	if len(staticPubBytes) != ed25519PubLen {
		return nil, fmt.Errorf("decrypted static key wrong size")
	}

	remoteStaticPub := ed25519.PublicKey(staticPubBytes)
	h.remoteStatic = remoteStaticPub

	// Convert remote Ed25519 to X25519
	remoteStaticX25519 := ed25519PublicToX25519(remoteStaticPub)

	// se: DH(responder_ephemeral, initiator_static_x25519)
	if h.localEphemeral == nil {
		return nil, fmt.Errorf("missing local ephemeral for se DH")
	}
	seShared, err := x25519DH(*h.localEphemeral, remoteStaticX25519)
	if err != nil {
		return nil, err
	}
	if err := h.mixKey(seShared); err != nil {
		return nil, err
	}

	// Mix in PAKE secret if present
	if h.pakeSecret != nil {
		if err := h.mixKey(*h.pakeSecret); err != nil {
			return nil, err
		}
	}

	// Decrypt payload
	encryptedPayload := msg3[offset:]
	if _, err := h.decryptAndHash(encryptedPayload); err != nil {
		return nil, err
	}

	// Derive session key
	sessionKey, err := h.deriveSessionKey()
	if err != nil {
		return nil, err
	}

	h.state = stateComplete
	return &StepOutput{
		Complete: &HandshakeResult{
			SessionKey:     sessionKey,
			RemoteStatic:   remoteStaticPub,
			TranscriptHash: h.handshakeHash,
		},
	}, nil
}

// --- Noise symmetric state operations ---

func (h *NoiseXXHandshake) mixKey(inputKeyMaterial [32]byte) error {
	output, err := HkdfSHA256(inputKeyMaterial[:], h.chainingKey[:], []byte(""), 64)
	if err != nil {
		return err
	}
	copy(h.chainingKey[:], output[:32])
	var derivedKey [32]byte
	copy(derivedKey[:], output[32:64])
	h.currentKey = &derivedKey
	return nil
}

func (h *NoiseXXHandshake) mixHash(data []byte) {
	hasher := sha256.New()
	hasher.Write(h.handshakeHash[:])
	hasher.Write(data)
	copy(h.handshakeHash[:], hasher.Sum(nil))
}

func (h *NoiseXXHandshake) encryptAndHash(plaintext []byte) ([]byte, error) {
	if h.currentKey == nil {
		return nil, fmt.Errorf("no encryption key available (mixKey not called)")
	}
	ciphertext, err := AeadEncrypt(CipherChaCha20Poly1305, *h.currentKey, zeroNonce, plaintext, h.handshakeHash[:])
	if err != nil {
		return nil, err
	}
	h.mixHash(ciphertext)
	return ciphertext, nil
}

func (h *NoiseXXHandshake) decryptAndHash(ciphertext []byte) ([]byte, error) {
	if h.currentKey == nil {
		return nil, fmt.Errorf("no decryption key available (mixKey not called)")
	}
	hBefore := h.handshakeHash
	h.mixHash(ciphertext)
	return AeadDecrypt(CipherChaCha20Poly1305, *h.currentKey, zeroNonce, ciphertext, hBefore[:])
}

func (h *NoiseXXHandshake) deriveSessionKey() ([32]byte, error) {
	out, err := HkdfSHA256(h.chainingKey[:], nil, HkdfInfoSessionKey, 32)
	if err != nil {
		return [32]byte{}, err
	}
	var key [32]byte
	copy(key[:], out)
	return key, nil
}

// --- Key conversion helpers ---

// ed25519PrivateToX25519 converts an Ed25519 private key to an X25519 private key.
// This mirrors the Rust `SigningKey.to_scalar_bytes()` operation.
func ed25519PrivateToX25519(priv ed25519.PrivateKey) [32]byte {
	// Ed25519 private key seed is the first 32 bytes. Hash it with SHA-512
	// and clamp to get the X25519 scalar, same as ed25519-dalek's to_scalar_bytes().
	h := sha512Sum(priv.Seed())
	// Clamp for X25519
	h[0] &= 248
	h[31] &= 127
	h[31] |= 64
	var result [32]byte
	copy(result[:], h[:32])
	return result
}

// sha512Sum computes SHA-512. Ed25519 key derivation uses SHA-512.
func sha512Sum(data []byte) [64]byte {
	return sha512.Sum512(data)
}

// ed25519PublicToX25519 converts an Ed25519 public key to an X25519 public key.
// This mirrors the Rust `VerifyingKey.to_montgomery()` operation.
func ed25519PublicToX25519(pub ed25519.PublicKey) [32]byte {
	// Use filippo.io/edwards25519 for the conversion
	p, err := new(edwards25519.Point).SetBytes(pub)
	if err != nil {
		// Should not happen with a valid Ed25519 public key
		var zero [32]byte
		return zero
	}
	montgomery := p.BytesMontgomery()
	var result [32]byte
	copy(result[:], montgomery)
	return result
}

// --- X25519 helpers ---

func generateX25519Keypair() (priv [32]byte, pub [32]byte, err error) {
	kp, err := GenerateX25519()
	if err != nil {
		return [32]byte{}, [32]byte{}, err
	}
	return kp.PrivateKeyBytes(), kp.PublicKeyBytes(), nil
}

func x25519DH(priv, peerPub [32]byte) ([32]byte, error) {
	shared, err := curve25519.X25519(priv[:], peerPub[:])
	if err != nil {
		return [32]byte{}, fmt.Errorf("X25519 DH failed: %w", err)
	}
	var result [32]byte
	copy(result[:], shared)
	return result, nil
}

// --- SAS Derivation ---

// NumericSAS derives a 6-digit numeric SAS from the handshake transcript hash.
func NumericSAS(transcriptHash [32]byte) (string, error) {
	derived, err := HkdfSHA256(transcriptHash[:], nil, HkdfInfoSAS, 4)
	if err != nil {
		return "", err
	}
	value := uint32(derived[0])<<24 | uint32(derived[1])<<16 | uint32(derived[2])<<8 | uint32(derived[3])
	value = value % 1_000_000
	return fmt.Sprintf("%06d", value), nil
}

// EmojiSAS derives a 4-emoji SAS from the handshake transcript hash.
func EmojiSAS(transcriptHash [32]byte) ([]string, error) {
	derived, err := HkdfSHA256(transcriptHash[:], nil, HkdfInfoSAS, 4)
	if err != nil {
		return nil, err
	}
	emojis := make([]string, 4)
	for i := 0; i < 4; i++ {
		emojis[i] = EmojiTable[derived[i]%64]
	}
	return emojis, nil
}
